// eas-cli/build/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserInputResourceClass = exports.BuildDistributionType = exports.BuildStatus = void 0;
var BuildStatus;
(function (BuildStatus) {
    BuildStatus["NEW"] = "new";
    BuildStatus["IN_QUEUE"] = "in-queue";
    BuildStatus["IN_PROGRESS"] = "in-progress";
    BuildStatus["ERRORED"] = "errored";
    BuildStatus["FINISHED"] = "finished";
    BuildStatus["CANCELED"] = "canceled";
})(BuildStatus = exports.BuildStatus || (exports.BuildStatus = {}));
var BuildDistributionType;
(function (BuildDistributionType) {
    BuildDistributionType["STORE"] = "store";
    BuildDistributionType["INTERNAL"] = "internal";
    BuildDistributionType["SIMULATOR"] = "simulator";
})(BuildDistributionType = exports.BuildDistributionType || (exports.BuildDistributionType = {}));
var UserInputResourceClass;
(function (UserInputResourceClass) {
    UserInputResourceClass["DEFAULT"] = "default";
    UserInputResourceClass["LARGE"] = "large";
    /**
     * @experimental
     * This resource class is not yet ready to be used in production. For testing purposes only. Might be deprecated / deleted at any time.
     */
    UserInputResourceClass["M1_EXPERIMENTAL"] = "m1-experimental";
})(UserInputResourceClass = exports.UserInputResourceClass || (exports.UserInputResourceClass = {}));
